internal/agent/device/dependency/dependency.go

```diff
@@ -1,20 +1,406 @@
+		if err != nil {
+			return err
+		}
+		return nil
+	case OCITypeArtifact:
+		_, err := m.podmanClient.PullArtifact(ctx, target, task.clientOps...)
+		if err != nil {
+			return err
+		}
+		return nil
+	default:
+		return fmt.Errorf("invalid oci type %s", ociType)
+	}
+}
+
+func (m *prefetchManager) setResult(image string, err error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	task := m.tasks[image]
+	task.err = err
+	task.done = true
+}
+
+func (m *prefetchManager) Schedule(ctx context.Context, targets []OCIPullTarget) error {
+	for _, target := range targets {
+		opts := []client.ClientOption{
+			client.Timeout(m.pullTimeout),
+		}
+		var cleanupFns []func()
+		if target.PullSecret != nil {
+			opts = append(opts, client.WithPullSecret(target.PullSecret.Path))
+			cleanupFns = append(cleanupFns, target.PullSecret.Cleanup)
+		}
+		if err := m.schedule(ctx, target.Reference, target.Type, cleanupFns, opts...); err != nil {
+			return fmt.Errorf("prefetch schedule: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (m *prefetchManager) schedule(ctx context.Context, image string, ociType OCIType, cleanupFns []func(), opts ...client.ClientOption) error {
+	m.mu.Lock()
+	if _, exists := m.tasks[image]; exists {
+		m.mu.Unlock()
+		return nil
+	}
+
+	if m.podmanClient.ImageExists(ctx, image) {
+		m.log.Debugf("Scheduled prefetch target already exists: %s", image)
+		// mark done for unified management flow
+		m.tasks[image] = &prefetchTask{
+			ociType:    ociType,
+			done:       true,
+			err:        nil,
+			cleanupFns: cleanupFns,
+		}
+		m.mu.Unlock()
+		return nil
+	}
+
+	// register task
+	task := &prefetchTask{
+		ociType:    ociType,
+		clientOps:  opts,
+		cleanupFns: cleanupFns,
+	}
+	m.tasks[image] = task
+	m.mu.Unlock()
+
+	timer := time.NewTimer(250 * time.Millisecond)
+	defer timer.Stop()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("failed to enqueue image %s: %w", image, ctx.Err())
+	case m.queue <- image:
+		return nil
+	case <-timer.C:
+		m.log.Warnf("Prefetch schedule failed for: %s: buffer full", image)
+		return fmt.Errorf("%w: buffer full", errors.ErrPrefetchNotReady)
+	}
+}
+
+func (m *prefetchManager) IsReady(ctx context.Context) bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if ctx.Err() != nil {
+		return false
+	}
+
+	for _, task := range m.tasks {
+		if !task.done || task.err != nil {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (m *prefetchManager) isTargetReady(image string) bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	task, ok := m.tasks[image]
+	if !ok {
+		return false
+	}
+	return task.err == nil && task.done
+}
+
+func (m *prefetchManager) Cleanup() {
+	m.log.Debug("Prefetch cleanup: canceling active task and draining queue")
+
+	m.mu.Lock()
+	for _, task := range m.tasks {
+		if task.cancelFn != nil {
+			task.cancelFn()
+		}
+		// fire cleanups
+		for _, cleanup := range task.cleanupFns {
+			cleanup()
+		}
+	}
+	m.mu.Unlock()
+
+	for {
+		select {
+		case <-m.queue:
```

Comment:
I _think_ there is a race condition here between the worker routine and Cleanup where the worker routine could set a cancelFunc on a task after we have already cleaned up the existing tasks. In the real flow I don't think it is likely to be hit.

Reply:
fixed thanks